File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.cs
using System;
using System.IO;

namespace AlgorithmStudy
{
    class boj9095
    {
        public static void Main(string[] args)
        {
            StreamReader sr = new StreamReader(new BufferedStream(Console.OpenStandardInput()));
            int[] factorials = new int[11];
            factorials[0] = 1;
            factorials[1] = 1;
            for (int i = 2; i < 11; i++)
                factorials[i] = i * factorials[i - 1];
            int t = int.Parse(sr.ReadLine());
            for (int i = 0; i < t; i++)
            {
                int[] numCount = new int[3];
                int n = int.Parse(sr.ReadLine());
                int temp = n;
                numCount[2] = temp / 3;

                int count = 0;
                for (int a = numCount[2]; a >= 0; a--)
                {
                    temp = n;
                    temp -= 3 * a;
                    numCount[1] = temp / 2;
                    temp %= 2;
                    numCount[0] = temp;
                    for (int b = numCount[1]; b >= 0; b--)
                    {
                        count += factorials[a + b + numCount[0]] / (factorials[a] * factorials[b] * factorials[numCount[0]]);
                        numCount[0] += 2;
                    }
                }

                Console.WriteLine(count);
            }
            sr.Close();
        }
    }
}

File: 백준/Silver/9095. 1， 2， 3 더하기/1， 2， 3 더하기.cc
#include <bits/stdc++.h>
using namespace std;

int main() {
	int fact[11];
	fact[0] = 1;
	fact[1] = 1;
	for (int i = 2; i <= 10; i++)
		fact[i] = fact[i - 1] * i;

	int t;
	cin >> t;
	for (int i = 0; i < t; i++) {
		int n;
		cin >> n;

		int count = 0;
		int numCount[3] = { n, 0, 0 };
		while (true) {
			int temp = fact[numCount[0] + numCount[1] + numCount[2]];
			temp /= fact[numCount[0]];
			temp /= fact[numCount[1]];
			temp /= fact[numCount[2]];
			count += temp;

			if (numCount[0] >= 2) {
				numCount[0] -= 2;
				numCount[1] += 1;
			}
			else if (numCount[0] < 2) {
				numCount[1] = 0;
				numCount[2] += 1;
				numCount[0] = n - numCount[2] * 3;
			}
			if (numCount[0] < 0)
				break;
		}

		printf("%d\n", count);
	}

	return 0;
}